### DISCUSSION TITLE
How to react to external resources outside of K8S

### DISCUSSION THREAD

**Discussion by AndreasBergmeier6176** [kubernetes-sigs/kubebuilder#3633] [Support] [answered]
We have a Cloud API with Resources/State that lives outside of K8S. Now we are trying to write a Controller that
1. Polls for all applicable Resources once every N seconds
2. Creates owned ReplicaSets for all these applicable Resources
3. Updates owned ReplicaSets for all these applicable Resources
4. Deletes owned ReplicaSets for all these applicable Resources
5. Watches owned ReplicaSets for all these applicable Resources

So far I could not find a guidance how best to tackle that with Kubebuilder.
Am I missing something?

**Comment by camilamacedo86** [comment, marked as answer]
Hi @AndreasBergmeier6176,

It's essential to fully understand the context to provide the best design solution. Additionally, it might be challenging for anyone here to come up with a domain-specific design solution. Controllers in Kubernetes primarily watch and observe changes in resources within the cluster, not outside. When a resource in the cluster is changed, created, or deleted, the reconciliation process is triggered. Ideally, your solution should run until the desired state declared in the reconciliation becomes the actual state.

However, you can explore options like:

**Using Go for HTTP Requests**: Leverage Go to make HTTP requests to any API/server. There are multiple approaches:
   - Your external server could create resources that represent their solution in the Kubernetes cluster.
   - You might implement a solution that polls the external server, perhaps using a cron job, to reflect changes within Kubernetes.

**Regarding Kubebuilder's Role**: Kubebuilder helps in creating projects, mainly following the Operator concept. It facilitates the process but requires understanding specific implementation details for unique scenarios.

Therefore, I recommend checking out [Kubebuilder's documentation](https://book.kubebuilder.io/getting-started) for getting started and understanding fundamental concepts.

Given the information provided, offering specific guidance is challenging. If you have more detailed questions about utilizing Kubebuilder for your specific use case, please feel free to open new issues or discussion topics. I hope you understand why I am closing this one, but feel free to re-open it if you see fit or to open new ones.

thank you !
